feat: add proof plan RPCs

tlovell-sxt opened the pull request:
# Rationale for this change

A barrier to usage of proof-of-sql against chain data is the creation of proof-of-sql proof plans. Proof plans are similar to normal sql logical plans, but they are specifically for proof-of-sql proving/verification. The input for creating a proof plan, at a high level, is query text and schema information for all tables in a query.

The chain has proof-of-sql schema information for tables inside queries, so it makes sense to provide proof plan APIs on the chain itself. This change adds a couple RPC calls for this - proofPlan and evmProofPlan. The former supports more queries while the latter can be verified in the evm.

As much of this logic as possible has been implemented in a stateless way, and all state is accessed via runtime APIs.

# What changes are included in this PR?
- **feat: add CommitmentId extension trait for commitments**
- **feat: add CommitmentScheme::into_any_concrete method**
- **feat: read commitments storage for multiple tables and any scheme**
- **feat: add runtime api for reading commitments storage for any scheme**
- **feat: add proof plan RPCs**
- **chore: increment runtime impl_version after proof-plan-related changes**

# Are these changes tested?
Yes.

AdaJane-SxT commented on rpc/src/commitments/limits.rs:
```diff
@@ -3,3 +3,18 @@ pub const PROOF_PLAN_SIZE_LIMIT: usize = 65_536;
 pub const NUM_TABLES_LIMIT: usize = 64;
+
+/// Input limit for sql query text.
+pub const QUERY_SIZE_LIMIT: usize = 65_536;
```
Does this also need to accommodate all the schemas of the tables?

tlovell-sxt replied:
I'm not sure what you mean by this

AdaJane-SxT commented on runtime/src/lib.rs:
```diff
@@ -160,7 +162,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     // This value is set to 100 to notify Polkadot-JS App (https://polkadot.js.org/apps) to use
     //   the compatible custom types.
     spec_version: 224,
-    impl_version: 1,
```
I think you need to increase `spec_version` here not `impl_version`.

tlovell-sxt replied:
I was actually reading into `RuntimeVersion`s and it seemed to me that if a change doesn't break consensus/transactions in any way, then it can be marked as an `impl_version` bump instead of a `spec_version`: https://docs.rs/sp-version/latest/sp_version/struct.RuntimeVersion.html#structfield.impl_version

Though, if you still disagree I don't mind changing it

AdaJane-SxT replied:
It looks like we've added some runtime apis in the pallet_commitments, which will need a runtime upgrade. When we perform a runtime upgrade, only the spec_version is checked. If the spec_version remains the same I believe the chain will see this runtime blob as identical in it's functionality and will not perform an upgrade. My understanding is that the impl version should be updated to denote changes in the underlying implementation of existing functionality. Such that runtimes sharing spec_versions but having different impl versions would accept the same inputs and return the same outputs for existing functionality, but the 'behind the scenes' may be slightly different.

zachfrederick commented on pallets/commitments/src/lib.rs:
```diff
@@ -299,4 +306,36 @@ pub mod pallet {
+        /// Returns `None` if no scheme has complete coverage of the provided tables.
+        pub fn table_commitments_any_scheme<'t>(
+            table_identifiers: impl IntoIterator<Item = &'t TableIdentifier> + Copy,
+        ) -> Option<AnyCommitmentScheme<ConcreteType<Vec<TableCommitmentBytes>>>> {
```
Possible nit: the return type for `table_commitments_any_scheme` is extra type-ey. 

We could update both return types to be an alias

```rust
// original 
pub type TableCommitments = Vec<TableCommitmentBytes>;
// updated
pub type MaybeTableCommitments = Option<TableCommitments>;



pub type SchemeTableCommitments = ConcreteType<Vec<TableCommitmentBytes>>;
pub type SchemeWrappedCommitments = AnyCommitmentScheme<SchemeTableCommitments>;
pub type MaybeSchemeCommitments = Option<SchemeWrappedCommitments>;
```

zachfrederick commented on rpc/src/commitments/api.rs:
```diff
@@ -38,4 +48,20 @@ pub trait CommitmentsApi<BH: Serialize> {
+
+    /// Returns proof plan for the given query text.
+    #[method(name = "commitments_v1_proofPlan", blocking)]
+    fn v1_proof_plan(
```
possible nit: Should the version come first in these names or last?

`fn v1_proof_plan -> fn proof_plan_v1`

And `commitments_v1_proofPlan -> commitments_proofPlan_v1`

tlovell-sxt replied:
I'm following the lead of existing RPC calls written by polkadot: https://github.com/paritytech/polkadot-sdk/blob/d99bafc738266f97bd03807909772b4471e56ce4/substrate/client/rpc-spec-v2/src/chain_head/api.rs#L148

tlovell-sxt commented:
pushed changes addressing feedback. So that the diff is easier to read for now, I just did a normal push, but after approval I'll perform this rebase to combine the commits:
```
pick 11581ee feat: add CommitmentId extension trait for commitments
pick 7c44ab7 feat: add CommitmentScheme::into_any_concrete method
pick f351ecf feat: read commitments storage for multiple tables and any scheme
fixup 0fd0ed9 refactor: type alias for table_commitments_any_scheme
pick b6fb900 feat: add runtime api for reading commitments storage for any scheme
fixup 21aed80 refactor: use type alias in runtime apis
pick fed72bc feat: add proof plan RPCs
fixup d828717 docs: add comment about normalization
fixup 0e55d80 fix: bump sxt-node package version to 1.1.0
pick 83d5ecb chore: increment runtime impl_version after proof-plan-related changes
```

tlovell-sxt commented:
updated spec_version instead of impl_version. Will perform this rebase after approval:
```
pick 11581ee feat: add CommitmentId extension trait for commitments
pick 7c44ab7 feat: add CommitmentScheme::into_any_concrete method
pick f351ecf feat: read commitments storage for multiple tables and any scheme
fixup 0fd0ed9 refactor: type alias for table_commitments_any_scheme
pick b6fb900 feat: add runtime api for reading commitments storage for any scheme
fixup 0a803fd fix: update spec_version instead of impl_version
pick fed72bc feat: add proof plan RPCs
fixup d828717 docs: add comment about normalization
fixup 0e55d80 fix: bump sxt-node package version to 1.1.0
reword 83d5ecb chore: increment runtime impl_version after proof-plan-related changes
fixup 21aed80 refactor: use type alias in runtime apis
```

AdaJane-SxT commented on Cargo.toml:
Ah, yeah apologies, it should be `node/Cargo.toml`
